### PR TITLE
ASU-1002: Add HASO lottery logic

### PR DESCRIPTION
**Pull request by jmp**
This PR adds HASO lottery logic to the backend.

Strictly speaking the HASO apartment distribution doesn't have "lottery", but I chose to call it lottery anyway to keep things consistent with the HITAS flow.

When an application is created, it is automatically added to a reservation queue for each apartment applied to in the application. This reservation queue works for both HITAS and HASO apartments, but in HASO case the applications in each queue are kept in an order based on their right of residence number. The lower the number, the higher precedence the application has.

When the application period has finished, we can call the `reserve_haso_apartment` function to declare the winner for a given apartment. Before declaring a winner, the current queue is persisted to database. The winning application for that apartment will be marked as `RESERVED`. If the same application applied to other apartments with a lower priority, those applications will be removed from their queues and marked as `REJECTED`. If there are multiple winning applications (with the same right of residence number) the applications with duplicate numbers are still treated as winners but marked as `REVIEW` so that they can be looked into in more detail.

Applications can be added to or removed from the queue manually with a comment. In such case (for example, removing an application who declined an apartment), a change event is recorded into the database.

Note, some parts of the data model are implemented slightly differently compared to the original diagrams, because I thought they made more sense this way. For example a "position change" integer does for the reservation queue change events doesn't really make sense when adding or removing an application from the queue.

### CODE REVIEW

**Comment by tomijarvi** [comment]
This needs a bit of time to review so I'll let you know when I have done a full review.

File: application_form/models.py [hunk 1]
```diff
@@ -98,3 +102,55 @@ class ApplicationApartment(models.Model):
         default=ApplicationState.SUBMITTED,
         verbose_name=_("application state"),
     )
+
+
+class ReservationQueue(models.Model):
```
Comment (tomijarvi) [inline application_form/models.py:107]: The name of this model could be better if it is `HousingQueue` (Asuntojono). The reservation state is one of the states in the item.

**Comment by jmp** [comment]
Clarified a couple of things with the PO:

* When an application is marked as RESERVED and then cancelled afterwards, the next applicant in line should be marked as RESERVED.
* When a late application is added to the queue, it is added last. However multiple late applications form their own "pool" and should be kept in order based on their right of residence number.

**Comment by jmp** [comment]
Made the (hopefully) final changes related to persisting the initial queue positions:

* Apartments are distributed at the same time for the whole project, not one apartment at a time.
* Before marking any applications as "RESERVED", "CANCELED", etc., the queue is persisted to database.
* Only after that the apartments are distributed to the applicants.

For this reason I changed the internal API a bit: instead of reserving apartments with `reserve_haso_apartment(apartment)`, you call the `distribute_haso_apartments(project)` function, which internally processes all the apartments belonging to that project.

**Review by frwickst — CHANGES_REQUESTED** [review]
Overall this shaping up nicely now. I have to say that reading the code is a joy as it is very well structured and easy to read, really good work @jmp !

Some minor notes that might be worth looking at still, nothing major.

File: application_form/services/queue.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        )
+        with transaction.atomic():
+            if application.type == ApplicationType.HASO:
+                # For HASO applications, the queue position is determined by the
+                # right of residence number.
+                position = _calculate_queue_position(queue, application_apartment)
+                _shift_queue_positions(queue, position)
+            else:
+                # HITAS applications always get added to the end of the queue
+                position = queue.queue_applications.count()
```
[the remark this answers is not in the document]
Comment (jmp) [reply application_form/services/queue.py:41]: This is a great suggestion. From the apartment lottery point of view, Hitas and Puolihitas are exactly the same, but this is not obvious at the moment.

File: application_form/services/queue.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+@transaction.atomic
+def remove_application_from_queue(
+    application_apartment: ApplicationApartment,
+    comment: str = "",
+) -> None:
+    """
+    Removes the application from the queue of the given apartment. This essentially
+    means that the application for this specific apartment was rejected, so the state
+    of the application for this apartment will be updated to "REJECTED".
```
[the remark this answers is not in the document]
Comment (jmp) [reply application_form/services/queue.py]: Good catch, fixed in the latest changeset.

**Review by frwickst — APPROVED** [review]
LGTM 👍

**Pull request merged**